### PR TITLE
Case 20585: Add missing Web Overlay functionality to Web Entities

### PR DESCRIPTION
**Pull request by SamGondelman**
https://highfidelity.manuscript.com/f/cases/20585/Add-missing-Web-Overlay-functionality-to-Web-Entities

Test plan:
- Run [this](https://gist.githubusercontent.com/SamGondelman/d4122d63b2f800f6e3e3fa741904c2fb/raw/6b82390cf7f13f01b18e5ae86e6d6b3144818a07/WebTest.js).  You'll see this:
image
- Press "o" several times.  The screens should all get higher resolution:
image
- Press "i" several times.  The resolution will go back down:
image
- Press "u".  All 10 pages will reload with a new URL:
image
- Press "u" again.  All 10 pages will now display a gif:
image
- Press "y".  The gifs should all run smoother.  Press it again and they'll all go back to a lower framerate.
- Verify that the tablet continues to work.  In particular: make sure you can still create entities with Create, and switch between domains with Goto.
- Run [this](https://gist.githubusercontent.com/SamGondelman/84d3b36c3fd0a29e1e2af105df23e33f/raw/9508e77c3e86d7876d3d5ddb650db7445727833d/WebEntityEventBridge.js).  You'll see this:
image
- When you press the buttons, colored cubes will appear in front of you:
image

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16444/HighFidelity-Beta-PR14711-879de1c.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16446/HighFidelity-Beta-PR14711-9803043.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16446/HighFidelity-Beta-PR14711-9803043.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16446/HighFidelity-Beta-Interface-PR14711-9803043.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16446/HighFidelity-Beta-PR14711-9803043.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16446/HighFidelity-Beta-Interface-PR14711-9803043.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16477/HighFidelity-Beta-PR14711-1d2270e.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16477/HighFidelity-Beta-PR14711-1d2270e.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16477/HighFidelity-Beta-Interface-PR14711-1d2270e.exe

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -3152,6 +3229,61 @@ void Application::onDesktopRootItemCreated(QQuickItem* rootItem) {
...
+    surfaceContext->setContextProperty("Quat", new Quat());
+    surfaceContext->setContextProperty("MyAvatar", DependencyManager::get<AvatarManager>()->getMyAvatar().get());
+    surfaceContext->setContextProperty("Entities", DependencyManager::get<EntityScriptingInterface>().data());
+    surfaceContext->setContextProperty("Snapshot", DependencyManager::get<Snapshot>().data());
+
+    if (setAdditionalContextProperties) {
+        auto tabletScriptingInterface = DependencyManager::get<TabletScriptingInterface>();
+        auto flags = tabletScriptingInterface->getFlags();
+
+        surfaceContext->setContextProperty("offscreenFlags", flags);
```
[the remark this answers is not in the document]
Comment (SamGondelman) [reply interface/src/Application.cpp]: yeah, my goal was to maintain the current behavior without questioning any of the bizarre choices too much, lest I lose my mind.

File: libraries/entities/src/EntityItem.h [hunk 2]
```diff
@@ -621,7 +630,10 @@ class EntityItem : public QObject, public SpatiallyNestable, public ReadWriteLoc
     float _angularDamping { ENTITY_ITEM_DEFAULT_ANGULAR_DAMPING };
     bool _visible { ENTITY_ITEM_DEFAULT_VISIBLE };
     bool _isVisibleInSecondaryCamera { ENTITY_ITEM_DEFAULT_VISIBLE_IN_SECONDARY_CAMERA };
+    RenderLayer _renderLayer { RenderLayer::WORLD };
+    PrimitiveMode _primitiveMode { PrimitiveMode::SOLID };
```
[the remark this answers is not in the document]
Comment (SamGondelman) [reply libraries/entities/src/EntityItem.h]: This PR includes the changes from this PR where these properties were added: https://github.com/highfidelity/hifi/pull/14667

And you're correct, they replace various overlay properties that all overlays have.

File: libraries/entities-renderer/src/RenderableWebEntityItem.h [hunk 3]
```diff
@@ -39,40 +56,59 @@ class WebEntityRenderer : public TypedEntityRenderer<WebEntityItem> {
...
+    QString _sourceURL;
+    uint16_t _dpi;
+    QString _scriptURL;
+    uint8_t _maxFPS;
+    WebInputMode _inputMode;
+
+    glm::vec3 _contextPosition;
+
     QTimer _timer;
     uint64_t _lastRenderTime { 0 };
```
[the remark this answers is not in the document]
Comment (SamGondelman) [reply libraries/entities-renderer/src/RenderableWebEntityItem.h]: I'm going to hold of on this now, but I've taken note of this for the next pass I do to clean up all the properties.

File: libraries/entities/src/WebEntityItem.h [hunk 4]
```diff
@@ -54,15 +52,38 @@ class WebEntityItem : public EntityItem {
...
+
 protected:
+    glm::u8vec3 _color;
+    float _alpha { 1.0f };
+
     QString _sourceUrl;
     uint16_t _dpi;
+    QString _scriptURL;
+    uint8_t _maxFPS;
+    WebInputMode _inputMode;
```
[the remark this answers is not in the document]
Comment (SamGondelman) [reply libraries/entities/src/WebEntityItem.h]: I'll do this later too, for all entity types.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16708/HighFidelity-Beta-PR14711-9a734b2.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16708/HighFidelity-Beta-PR14711-9a734b2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16708/HighFidelity-Beta-Interface-PR14711-9a734b2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16708/HighFidelity-Beta-PR14711-9a734b2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16708/HighFidelity-Beta-Interface-PR14711-9a734b2.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16729/HighFidelity-Beta-PR14711-24b62ca.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16729/HighFidelity-Beta-PR14711-24b62ca.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/16729/HighFidelity-Beta-Interface-PR14711-24b62ca.dmg

File: interface/src/Application.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jherico) [reply interface/src/Application.cpp]: The point of the capitalization is that QML won't let you capitalize property names....  so in the QML code if you see a var name with a capital starting letter you know it's not a property, but a global.  `offscreenFlags` was probably just missed.

**Comment by qaAdamSmith82** [comment]
Currently Testing

**Pull request merged**